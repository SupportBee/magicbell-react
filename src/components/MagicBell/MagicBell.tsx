import { useMagicBellEvent } from '@magicbell/react-headless';
import INotificationStore from '@magicbell/react-headless/dist/types/INotificationStore';
import IRemoteNotification from '@magicbell/react-headless/dist/types/IRemoteNotification';
import React, { useRef } from 'react';
import useToggle from 'react-use/lib/useToggle';
import { IMagicBellTheme } from '../../context/Theme';
import { CustomLocale } from '../../lib/i18n';
import { DeepPartial } from '../../lib/types';
import Bell from '../Bell';
import MagicBellProvider from '../MagicBellProvider';

type StoreConfig = {
  id: string;
  defaultQueryParams: Object;
  defaults?: Partial<Omit<INotificationStore, 'context'>>;
};

export interface Props {
  apiKey: string;
  userEmail?: string;
  userExternalId?: string;
  userKey?: string;
  children: (params: {
    launcherRef: React.RefObject<Element>;
    isOpen: boolean;
    toggle: () => void;
  }) => JSX.Element;
  theme?: DeepPartial<IMagicBellTheme>;
  BellIcon?: JSX.Element;
  defaultIsOpen?: boolean;
  stores?: StoreConfig[];
  locale?: string | CustomLocale;
  images?: Partial<{
    emptyInboxUrl: string;
  }>;
  serverURL?: string;
  onNewNotification?: (notification: IRemoteNotification) => void;
  onToggle?: (isOpen: boolean) => void;
  bellCounter?: 'unread' | 'unseen';
}

/**
 * Magicbell root component. Use this one in your application.
 *
 * @param apiKey API key of the MagicBell project
 * @param userEmail Email of the user whose notifications will be displayed
 * @param userExternalId External ID of the user whose notifications will be displayed
 * @param userKey Computed HMAC of the user whose notifications will be displayed, compute this with the secret of the magicbell project
 * @param theme Object to customize the theme
 * @param BellIcon Icon for the bell
 * @param defaultIsOpen Show the children when the component is rendered. It is false by default.
 * @param stores Configuration of stores to be created
 * @param locale Locale to use in the components
 * @param onNewNotification Function called when a notification is created.
 * @param onToggle Function called when the bell is clicked.
 * @param bellCounter Counter to show in the bell. If set to 'unread' it will show the number of unread notifications.
 *
 * @example
 * <MagicBell
 *   apiKey={MAGICBELL_API_KEY}
 *   userEmail={email}
 *   BellIcon={<MyIcon />}
 * >
 *   {(props) => <NotificationInbox {...props} />}
 * </MagicBell>
 */
export default function MagicBell({
  children,
  BellIcon,
  defaultIsOpen = false,
  onNewNotification,
  onToggle,
  bellCounter = 'unseen',
  ...settings
}: Props) {
  const launcherRef = useRef(null);
  const [isOpen, toggleChildren] = useToggle(defaultIsOpen);

  const handleToggle = () => {
    toggleChildren();
    onToggle?.(isOpen);
  };

  const handleNewNotification = (notification: IRemoteNotification) => {
    onNewNotification?.(notification);
  };

  useMagicBellEvent('notifications.new', handleNewNotification);

  return (
    <MagicBellProvider {...settings}>
      <div>
        <div ref={launcherRef} aria-expanded={isOpen}>
          <Bell onClick={handleToggle} Icon={BellIcon} counter={bellCounter} />
        </div>
        {isOpen && children({ isOpen, toggle: handleToggle, launcherRef })}
      </div>
    </MagicBellProvider>
  );
}
