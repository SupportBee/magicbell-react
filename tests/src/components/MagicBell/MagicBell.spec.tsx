import { pushEventAggregator } from '@magicbell/react-headless';
import { fireEvent, render, RenderResult, screen, waitFor } from '@testing-library/react';
import faker from 'faker';
import { Response, Server } from 'miragejs';
import React from 'react';
import MagicBell, { Header } from '../../../../src';
import NotificationFactory from '../../../../tests/factories/NotificationFactory';
import { sampleConfig } from '../../../factories/ConfigFactory';

describe('components', () => {
  describe('MagicBell', () => {
    const apiKey = faker.random.alphaNumeric(10);
    const userEmail = faker.internet.email();
    const userKey = faker.random.alphaNumeric(10);

    let server;
    let view: RenderResult;

    beforeEach(() => {
      server = new Server({
        environment: 'test',
        urlPrefix: 'https://api.magicbell.com',
        trackRequests: true,
        timing: 50,
      });
      server.get('/notifications', {
        total: 5,
        current_page: 1,
        per_page: 15,
        total_pages: 1,
        project_id: 7,
        unseen_count: 0,
        unread_count: 4,
        notifications: NotificationFactory.buildList(5),
      });
      server.get('/config', sampleConfig);
      server.get('/notification_preferences', {
        notification_preferences: {
          categories: {
            comments: { email: false },
          },
        },
      });
      server.post('/notifications/seen', new Response(204, {}, ''));

      view = render(
        <MagicBell apiKey={apiKey} userEmail={userEmail} userKey={userKey}>
          {() => <div data-testid="children" />}
        </MagicBell>,
      );
    });

    afterEach(() => {
      view.unmount();
      server.shutdown();
    });

    describe('render', () => {
      it('renders a bell only', async () => {
        await waitFor(() => expect(view.container).toMatchSnapshot());
      });

      describe('the bell is clicked', () => {
        it('renders the children', () => {
          fireEvent.click(view.getByTestId('bell'));
          expect(view.container).toMatchSnapshot();
        });
      });

      describe('a custom bell icon is given', () => {
        it('renders the custom icon', async () => {
          const BellIcon = () => <p>Custom bell icon</p>;
          view.rerender(
            <MagicBell
              apiKey={apiKey}
              userEmail={userEmail}
              userKey={userKey}
              BellIcon={<BellIcon />}
            >
              {() => <div data-testid="children" />}
            </MagicBell>,
          );

          await waitFor(() => expect(view.container).toMatchSnapshot());
        });
      });

      describe('it is open by default', () => {
        it('renders the children when it is mounted', () => {
          const { unmount } = render(
            <MagicBell apiKey={apiKey} userEmail={userEmail} userKey={userKey} defaultIsOpen>
              {() => <div data-testid="children" />}
            </MagicBell>,
          );

          expect(screen.queryByTestId('children')).toBeInTheDocument();
          unmount();
        });
      });

      describe('custom locale', () => {
        it('sets the custom translations', async () => {
          const { unmount } = render(
            <MagicBell
              apiKey={apiKey}
              userEmail={userEmail}
              userKey={userKey}
              locale={{
                name: 'en-US',
                translations: {
                  header: { 'mark-all-read': 'Please mark all as read' },
                },
              }}
              defaultIsOpen
            >
              {() => <Header onAllRead={jest.fn()} />}
            </MagicBell>,
          );

          expect(await screen.findByTestId('mark-all-as-read')).toMatchSnapshot();
          unmount();
        });
      });

      describe('show unread notifications', () => {
        it('shows the number of unread notifications', async () => {
          view.rerender(
            <MagicBell apiKey={apiKey} userEmail={userEmail} userKey={userKey} bellCounter="unread">
              {() => <div data-testid="children" />}
            </MagicBell>,
          );

          const bell = await view.findByTestId('bell');
          await waitFor(() => expect(bell).toHaveTextContent('4'));
        });
      });
    });

    describe('.toggleNotificationInbox', () => {
      it('toggles the children', async () => {
        const { unmount } = render(
          <MagicBell apiKey={apiKey} userEmail={userEmail} userKey={userKey} defaultIsOpen>
            {({ toggle }) => <div data-testid="children" onClick={toggle} />}
          </MagicBell>,
        );
        fireEvent.click(await screen.getByTestId('children'));

        expect(view.queryByTestId('children')).not.toBeInTheDocument();
        unmount();
      });

      it('calls the onToggle callback', async () => {
        const onToggle = jest.fn();
        const { unmount } = render(
          <MagicBell
            apiKey={apiKey}
            userEmail={userEmail}
            userKey={userKey}
            onToggle={onToggle}
            defaultIsOpen
          >
            {({ toggle }) => <div data-testid="children" onClick={toggle} />}
          </MagicBell>,
        );
        fireEvent.click(await screen.getByTestId('children'));

        expect(onToggle).toHaveBeenCalledTimes(1);
        unmount();
      });
    });

    describe('component did mount', () => {
      it('sets the headers for fetching from the API', () => {
        const { unmount } = render(
          <MagicBell apiKey={apiKey} userEmail={userEmail} userKey={userKey}>
            {() => <div data-testid="children" />}
          </MagicBell>,
        );

        const requests = server.pretender.handledRequests;
        expect(requests[0].requestHeaders).toMatchObject({
          'X-MAGICBELL-API-KEY': apiKey,
          'X-MAGICBELL-USER-EMAIL': userEmail,
          'X-MAGICBELL-USER-HMAC': userKey,
        });

        unmount();
      });

      describe('with external id', () => {
        it('sets the external id header for etching from the API', async () => {
          const userExternalId = faker.random.alphaNumeric(15);

          const { unmount } = render(
            <MagicBell apiKey={apiKey} userExternalId={userExternalId} userKey={userKey}>
              {() => <div data-testid="children" />}
            </MagicBell>,
          );

          await waitFor(() => expect(server.pretender.handledRequests.length).toBe(2));
          const requests = server.pretender.handledRequests;
          expect(requests[1].requestHeaders).toMatchObject({
            'X-MAGICBELL-API-KEY': apiKey,
            'X-MAGICBELL-USER-EXTERNAL-ID': userExternalId,
            'X-MAGICBELL-USER-HMAC': userKey,
          });

          unmount();
        });
      });
    });

    describe('.handleToggle', () => {
      describe('there is an onToggle callback defined', () => {
        let onToggle;

        beforeEach(() => {
          onToggle = jest.fn();
          view.rerender(
            <MagicBell
              apiKey={apiKey}
              userEmail={userEmail}
              userKey={userKey}
              onToggle={onToggle}
              defaultIsOpen
            >
              {() => <div data-testid="children" />}
            </MagicBell>,
          );
        });

        it('calls the onToggle callback', () => {
          fireEvent.click(view.getByTestId('bell'));

          expect(onToggle).toHaveBeenCalledTimes(1);
          expect(onToggle).toHaveBeenCalledWith(false);
        });
      });
    });

    describe('.handleNewNotification', () => {
      it('calls the onNewNotification callback when a new notification is received', () => {
        const onNewNotification = jest.fn();
        const { unmount } = render(
          <MagicBell
            apiKey={apiKey}
            userEmail={userEmail}
            userKey={userKey}
            onNewNotification={onNewNotification}
          >
            {() => <div data-testid="children" />}
          </MagicBell>,
        );

        const notification = NotificationFactory.build();
        pushEventAggregator.emit('notifications.new', notification);

        expect(onNewNotification).toHaveBeenCalledTimes(1);
        expect(onNewNotification).toHaveBeenCalledWith(notification);

        unmount();
      });
    });
  });
});
