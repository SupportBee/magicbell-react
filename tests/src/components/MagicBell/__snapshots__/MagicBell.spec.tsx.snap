// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components MagicBell render a custom bell icon is given renders the custom icon 1`] = `
.emotion-0 {
  -webkit-backface-visibility: visible;
  backface-visibility: visible;
  -webkit-background-clip: border-box;
  background-clip: border-box;
  background-color: transparent;
  background-origin: padding-box;
  -webkit-background-position: 0 0;
  background-position: 0 0;
  -webkit-background-size: auto;
  background-size: auto;
  border-radius: 0;
  border: 0;
  box-shadow: none;
  box-sizing: content-box;
  clear: none;
  clip: auto;
  color: inherit;
  cursor: auto;
  direction: inherit;
  display: block;
  left: auto;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  font-size: 14px;
  font-weight: normal;
  height: auto;
  -webkit-hyphens: manual;
  -moz-hyphens: manual;
  -ms-hyphens: manual;
  hyphens: manual;
  letter-spacing: normal;
  line-height: 1;
  margin: 0;
  max-height: none;
  max-width: none;
  min-height: 0;
  min-width: 0;
  opacity: 1;
  outline: invert none medium;
  padding: 0;
  perspective: none;
  position: static;
  right: auto;
  top: auto;
  -webkit-transform: none;
  -moz-transform: none;
  -ms-transform: none;
  transform: none;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  visibility: inherit;
  white-space: normal;
  width: auto;
  word-spacing: normal;
  z-index: auto;
  display: block;
  cursor: pointer;
  position: relative!important;
  width: 24px!important;
}

.emotion-0 div,
.emotion-0 span,
.emotion-0 p,
.emotion-0 a,
.emotion-0 strong,
.emotion-0 b,
.emotion-0 u,
.emotion-0 i,
.emotion-0 ol,
.emotion-0 ul,
.emotion-0 li,
.emotion-0 form,
.emotion-0 label,
.emotion-0 table,
.emotion-0 tr,
.emotion-0 th,
.emotion-0 td,
.emotion-0 article,
.emotion-0 aside,
.emotion-0 footer,
.emotion-0 header,
.emotion-0 nav,
.emotion-0 section {
  margin: 0;
  padding: 0;
  border: 0;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
}

.emotion-0 a {
  cursor: pointer!important;
  color: inherit!important;
}

.emotion-0 button {
  background: transparent;
  border: none;
  font-size: inherit!important;
  color: inherit;
  padding: 0;
  cursor: pointer!important;
}

.emotion-1 {
  position: relative!important;
}

.emotion-1>* {
  height: 100%;
  width: 100%;
}

<div>
  <div>
    <div
      aria-expanded="false"
    >
      <a
        class="emotion-0"
        data-testid="bell"
      >
        <div
          class="emotion-1"
        >
          <p>
            Custom bell icon
          </p>
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`components MagicBell render custom locale sets the custom translations 1`] = `
<button
  data-testid="mark-all-as-read"
>
  Please mark all as read
</button>
`;

exports[`components MagicBell render renders a bell only 1`] = `
.emotion-0 {
  -webkit-backface-visibility: visible;
  backface-visibility: visible;
  -webkit-background-clip: border-box;
  background-clip: border-box;
  background-color: transparent;
  background-origin: padding-box;
  -webkit-background-position: 0 0;
  background-position: 0 0;
  -webkit-background-size: auto;
  background-size: auto;
  border-radius: 0;
  border: 0;
  box-shadow: none;
  box-sizing: content-box;
  clear: none;
  clip: auto;
  color: inherit;
  cursor: auto;
  direction: inherit;
  display: block;
  left: auto;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  font-size: 14px;
  font-weight: normal;
  height: auto;
  -webkit-hyphens: manual;
  -moz-hyphens: manual;
  -ms-hyphens: manual;
  hyphens: manual;
  letter-spacing: normal;
  line-height: 1;
  margin: 0;
  max-height: none;
  max-width: none;
  min-height: 0;
  min-width: 0;
  opacity: 1;
  outline: invert none medium;
  padding: 0;
  perspective: none;
  position: static;
  right: auto;
  top: auto;
  -webkit-transform: none;
  -moz-transform: none;
  -ms-transform: none;
  transform: none;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  visibility: inherit;
  white-space: normal;
  width: auto;
  word-spacing: normal;
  z-index: auto;
  display: block;
  cursor: pointer;
  position: relative!important;
  width: 24px!important;
}

.emotion-0 div,
.emotion-0 span,
.emotion-0 p,
.emotion-0 a,
.emotion-0 strong,
.emotion-0 b,
.emotion-0 u,
.emotion-0 i,
.emotion-0 ol,
.emotion-0 ul,
.emotion-0 li,
.emotion-0 form,
.emotion-0 label,
.emotion-0 table,
.emotion-0 tr,
.emotion-0 th,
.emotion-0 td,
.emotion-0 article,
.emotion-0 aside,
.emotion-0 footer,
.emotion-0 header,
.emotion-0 nav,
.emotion-0 section {
  margin: 0;
  padding: 0;
  border: 0;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
}

.emotion-0 a {
  cursor: pointer!important;
  color: inherit!important;
}

.emotion-0 button {
  background: transparent;
  border: none;
  font-size: inherit!important;
  color: inherit;
  padding: 0;
  cursor: pointer!important;
}

.emotion-1 {
  position: relative!important;
}

.emotion-1>* {
  height: 100%;
  width: 100%;
}

<div>
  <div>
    <div
      aria-expanded="false"
    >
      <a
        class="emotion-0"
        data-testid="bell"
      >
        <div
          class="emotion-1"
        >
          <svg
            fill="#3498F4"
            height="32"
            viewBox="0 0 32 32"
            width="32"
          >
            <path
              d="M15.9997 4.70663C14.3971 4.71182 12.8617 5.34576 11.7284 6.47013C10.5951 7.59449 9.95592 9.11801 9.95038 10.7082V15.0907C9.95038 16.4805 9.85234 17.3932 9.62941 18.1738C9.39598 18.9845 9.04584 19.5809 8.55447 20.3349C8.38874 20.5862 8.23117 20.8144 8.08761 21.0217C7.66277 21.6158 8.58949 21.7583 9.06218 21.8324C10.8129 22.115 13.2639 22.2957 16.0102 22.2957C18.7565 22.2957 21.2052 22.1196 22.9582 21.8324C24.1791 21.6344 24.0822 21.3031 23.5897 20.5584C23.0819 19.7929 22.7038 19.1791 22.44 18.3718C22.1762 17.5646 22.0467 16.6496 22.0467 15.3119V10.7082C22.0442 9.11807 21.4069 7.59366 20.2742 6.4688C19.1414 5.34394 17.6057 4.71031 16.0032 4.70663H15.9997ZM16.0032 2C18.3263 2.01068 20.551 2.93195 22.1926 4.56308C23.8342 6.19422 24.7596 8.40302 24.7673 10.7082V15.3166C24.7673 16.3589 24.8502 17.0214 25.0217 17.5414C25.1933 18.0614 25.4711 18.5085 25.8528 19.0829C26.1294 19.501 26.4048 19.8647 26.6289 20.1623C27.1588 20.8642 27.5136 21.3344 27.4996 21.976V22.0061C27.4996 23.2234 25.9263 24.0908 23.3854 24.4985C21.4713 24.8089 18.858 25 16.0009 25C13.1437 25 10.5304 24.8089 8.61516 24.4985C6.07428 24.0874 4.50214 23.2245 4.50214 22.0061V21.9957C4.46829 21.3286 4.83944 20.8306 5.40434 20.0766C5.64944 19.7489 5.95057 19.3447 6.26686 18.8629C6.61701 18.3267 6.86094 17.9201 7.001 17.4325C7.15039 16.9091 7.21575 16.2153 7.21575 15.0884V10.7082C7.22192 8.40035 8.14864 6.18875 9.79334 4.55693C11.438 2.92511 13.6669 2.00581 15.9927 2H16.0032Z"
            />
            <path
              d="M20.4988 26C20.4489 27.0446 19.9795 28.0338 19.1834 28.7721C18.7655 29.1614 18.2692 29.4702 17.7228 29.6809C17.1765 29.8916 16.5909 30 15.9994 30C15.408 30 14.8223 29.8916 14.276 29.6809C13.7296 29.4702 13.2333 29.1614 12.8154 28.7721C12.0193 28.0338 11.5499 27.0446 11.5 26C12.3651 26.0703 13.3047 26.1198 14.2916 26.1473C14.3561 26.4337 14.5025 26.6985 14.7154 26.9139C14.9283 27.1293 15.1998 27.2872 15.5013 27.371C15.8027 27.4548 16.1228 27.4613 16.4279 27.3898C16.733 27.3184 17.0116 27.1716 17.2345 26.9651C17.4739 26.7413 17.6382 26.4575 17.7072 26.1484C18.6941 26.1209 19.6337 26.0714 20.5 26.0011"
            />
          </svg>
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`components MagicBell render the bell is clicked renders the children 1`] = `
.emotion-0 {
  -webkit-backface-visibility: visible;
  backface-visibility: visible;
  -webkit-background-clip: border-box;
  background-clip: border-box;
  background-color: transparent;
  background-origin: padding-box;
  -webkit-background-position: 0 0;
  background-position: 0 0;
  -webkit-background-size: auto;
  background-size: auto;
  border-radius: 0;
  border: 0;
  box-shadow: none;
  box-sizing: content-box;
  clear: none;
  clip: auto;
  color: inherit;
  cursor: auto;
  direction: inherit;
  display: block;
  left: auto;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  font-size: 14px;
  font-weight: normal;
  height: auto;
  -webkit-hyphens: manual;
  -moz-hyphens: manual;
  -ms-hyphens: manual;
  hyphens: manual;
  letter-spacing: normal;
  line-height: 1;
  margin: 0;
  max-height: none;
  max-width: none;
  min-height: 0;
  min-width: 0;
  opacity: 1;
  outline: invert none medium;
  padding: 0;
  perspective: none;
  position: static;
  right: auto;
  top: auto;
  -webkit-transform: none;
  -moz-transform: none;
  -ms-transform: none;
  transform: none;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  visibility: inherit;
  white-space: normal;
  width: auto;
  word-spacing: normal;
  z-index: auto;
  display: block;
  cursor: pointer;
  position: relative!important;
  width: 24px!important;
}

.emotion-0 div,
.emotion-0 span,
.emotion-0 p,
.emotion-0 a,
.emotion-0 strong,
.emotion-0 b,
.emotion-0 u,
.emotion-0 i,
.emotion-0 ol,
.emotion-0 ul,
.emotion-0 li,
.emotion-0 form,
.emotion-0 label,
.emotion-0 table,
.emotion-0 tr,
.emotion-0 th,
.emotion-0 td,
.emotion-0 article,
.emotion-0 aside,
.emotion-0 footer,
.emotion-0 header,
.emotion-0 nav,
.emotion-0 section {
  margin: 0;
  padding: 0;
  border: 0;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
}

.emotion-0 a {
  cursor: pointer!important;
  color: inherit!important;
}

.emotion-0 button {
  background: transparent;
  border: none;
  font-size: inherit!important;
  color: inherit;
  padding: 0;
  cursor: pointer!important;
}

.emotion-1 {
  position: relative!important;
}

.emotion-1>* {
  height: 100%;
  width: 100%;
}

<div>
  <div>
    <div
      aria-expanded="true"
    >
      <a
        class="emotion-0"
        data-testid="bell"
      >
        <div
          class="emotion-1"
        >
          <svg
            fill="#3498F4"
            height="32"
            viewBox="0 0 32 32"
            width="32"
          >
            <path
              d="M15.9997 4.70663C14.3971 4.71182 12.8617 5.34576 11.7284 6.47013C10.5951 7.59449 9.95592 9.11801 9.95038 10.7082V15.0907C9.95038 16.4805 9.85234 17.3932 9.62941 18.1738C9.39598 18.9845 9.04584 19.5809 8.55447 20.3349C8.38874 20.5862 8.23117 20.8144 8.08761 21.0217C7.66277 21.6158 8.58949 21.7583 9.06218 21.8324C10.8129 22.115 13.2639 22.2957 16.0102 22.2957C18.7565 22.2957 21.2052 22.1196 22.9582 21.8324C24.1791 21.6344 24.0822 21.3031 23.5897 20.5584C23.0819 19.7929 22.7038 19.1791 22.44 18.3718C22.1762 17.5646 22.0467 16.6496 22.0467 15.3119V10.7082C22.0442 9.11807 21.4069 7.59366 20.2742 6.4688C19.1414 5.34394 17.6057 4.71031 16.0032 4.70663H15.9997ZM16.0032 2C18.3263 2.01068 20.551 2.93195 22.1926 4.56308C23.8342 6.19422 24.7596 8.40302 24.7673 10.7082V15.3166C24.7673 16.3589 24.8502 17.0214 25.0217 17.5414C25.1933 18.0614 25.4711 18.5085 25.8528 19.0829C26.1294 19.501 26.4048 19.8647 26.6289 20.1623C27.1588 20.8642 27.5136 21.3344 27.4996 21.976V22.0061C27.4996 23.2234 25.9263 24.0908 23.3854 24.4985C21.4713 24.8089 18.858 25 16.0009 25C13.1437 25 10.5304 24.8089 8.61516 24.4985C6.07428 24.0874 4.50214 23.2245 4.50214 22.0061V21.9957C4.46829 21.3286 4.83944 20.8306 5.40434 20.0766C5.64944 19.7489 5.95057 19.3447 6.26686 18.8629C6.61701 18.3267 6.86094 17.9201 7.001 17.4325C7.15039 16.9091 7.21575 16.2153 7.21575 15.0884V10.7082C7.22192 8.40035 8.14864 6.18875 9.79334 4.55693C11.438 2.92511 13.6669 2.00581 15.9927 2H16.0032Z"
            />
            <path
              d="M20.4988 26C20.4489 27.0446 19.9795 28.0338 19.1834 28.7721C18.7655 29.1614 18.2692 29.4702 17.7228 29.6809C17.1765 29.8916 16.5909 30 15.9994 30C15.408 30 14.8223 29.8916 14.276 29.6809C13.7296 29.4702 13.2333 29.1614 12.8154 28.7721C12.0193 28.0338 11.5499 27.0446 11.5 26C12.3651 26.0703 13.3047 26.1198 14.2916 26.1473C14.3561 26.4337 14.5025 26.6985 14.7154 26.9139C14.9283 27.1293 15.1998 27.2872 15.5013 27.371C15.8027 27.4548 16.1228 27.4613 16.4279 27.3898C16.733 27.3184 17.0116 27.1716 17.2345 26.9651C17.4739 26.7413 17.6382 26.4575 17.7072 26.1484C18.6941 26.1209 19.6337 26.0714 20.5 26.0011"
            />
          </svg>
        </div>
      </a>
    </div>
    <div
      data-testid="children"
    />
  </div>
</div>
`;
